clippy_lints/src/borrow_deref_ref.rs

```diff
@@ -0,0 +1,118 @@
+use crate::reference::DEREF_ADDROF;
+use clippy_utils::diagnostics::span_lint_and_then;
+use clippy_utils::source::snippet_opt;
+use clippy_utils::ty::implements_trait;
+use clippy_utils::{get_parent_expr, is_lint_allowed};
+use rustc_errors::Applicability;
+use rustc_hir::{ExprKind, UnOp};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_middle::mir::Mutability;
+use rustc_middle::ty;
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for `&*(&T)`.
+    ///
+    /// ### Why is this bad?
+    /// When people deref on an immutable reference `&T`, they may expect return `&U`.
+    /// Accutually `&* (&T)` is still `&T`.
+    /// if you want to deref explicitly, `&** (&T)` is what you need.
+    /// If you want to reborrow, `&T` is enough (`&T` is Copy).
+    ///
+    /// ### Known problems
+    /// false negative on such code:
+    /// ```
+    /// let x = &12;
+    /// let addr_x = &x as *const _ as usize;
+    /// let addr_y = &&*x as *const _ as usize; // assert ok now, and lint triggerd.
+    ///                                         // But if we fix it, assert will fail.
+    /// assert_ne!(addr_x, addr_y);
+    /// ```
+    ///
+    /// ### Example
+    /// ```rust
+    /// let s = &String::new();
+    ///
+    /// // Bad
+    /// let a: &String = &* s;
+    /// foo(&*s);
+    ///
+    /// // Good
+    /// let a: &String = s;
+    /// foo(&**s);
+    ///
+    /// fn foo(_: &str){ }
+    /// ```
+    #[clippy::version = "1.59.0"]
+    pub BORROW_DEREF_REF,
+    complexity,
+    "deref on an immutable reference returns the same type as itself"
+}
+
+declare_lint_pass!(BorrowDerefRef => [BORROW_DEREF_REF]);
+
+impl LateLintPass<'_> for BorrowDerefRef {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, e: &'tcx rustc_hir::Expr<'_>) {
+        if_chain! {
+            if !e.span.from_expansion();
+            if let ExprKind::AddrOf(_, Mutability::Not, addrof_target) = e.kind;
+            if !addrof_target.span.from_expansion();
+            if let ExprKind::Unary(UnOp::Deref, deref_target) = addrof_target.kind;
+            if !deref_target.span.from_expansion();
+            if !matches!(deref_target.kind, ExprKind::Unary(UnOp::Deref, ..) );
+            let ref_ty = cx.typeck_results().expr_ty(deref_target);
+            if let ty::Ref(_, inner_ty, Mutability::Not) = ref_ty.kind();
+            then{
+
+                if let Some(parent_expr) = get_parent_expr(cx, e){
+                    if matches!(deref_target.kind, ExprKind::Path(..) | ExprKind::Field(..)) {
+                        if matches!(parent_expr.kind, ExprKind::AddrOf(_, Mutability::Mut, _)) {
```

Comment:
This needs a check for auto-borrow as well. Something like `cx.typeck_results().expr_adjustments(e).first().map_or(false, |a| matches!(a.kind, Adjust::Borrow(AutoBorrow::Ref(_, AutoRefMutability::Mut {..}))))`

Reply:
I guess `if !e.span.from_expansion();` excludes this case. (I'm not sure)

Reply:
No. Auto-borrow comes from the receiver in a method call. e.g. `(&*x).foo()` can borrow `&*x`

Reply:
I'm not sure I understand your meaning correctly.
Do you mean an `&mut` is auto-adjusted to `&*_` , and output `&mut &* _` ?

Reply:
When calling a method taking `&self` or `&mut self` the compiler will insert a borrow to make the type match if needed. For references this can only happen when a trait is implemented on a method (e.g. `io::Read` is implemented on `&[u8]).
A read example would be `(&*bytes).read(count)` which is intended to leave `bytes` unmodified after calling `read`.

Reply:
Here's a minimal example (can be a test). The `&*` cannot be removed because it would make the auto borrow not possible.

```rust
let src = &[0][..];
(&*src).read_to_end(&mut vec![]);
```

Reply:
Looks like this is fixed now.